README.md
# dynamic_theming

A new Flutter package project.

## Getting Started

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## [1.0.0] - 06/22/2020

* Initial release.


test/dynamic_theming_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:dynamic_theming/dynamic_theming.dart';

void main() {
  test('', () {
    
  });
}


lib/dynamic_theming.dart
library dynamic_theming;

import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

import 'package:dynamic_theming/src/theme_controller.dart';
export 'package:dynamic_theming/src/theme_controller.dart';

class DynamicThemedApp extends StatefulWidget {
  final ThemeController controller;
  final String title;
  final Widget home;
  final bool automaticSystemBars;
  final Color statusBarColor;

  DynamicThemedApp({
    this.controller,
    @required this.title,
    @required this.home,
    this.automaticSystemBars = false,
    this.statusBarColor = Colors.transparent,
  })  : assert(title != null),
        assert(home != null);

  @override
  _DynamicThemedAppState createState() => _DynamicThemedAppState();
}

class _DynamicThemedAppState extends State<DynamicThemedApp> with WidgetsBindingObserver {
  ThemeController theme;
  
  @override
  void didChangePlatformBrightness() {
    super.didChangePlatformBrightness();
    if (theme != null &&
        widget.automaticSystemBars &&
        theme.systemThemeEnabled)
      theme.setSystemBarsColor(
          theme.currentBrightness, widget.statusBarColor);
  }

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addObserver(this);
  }

  @override
  void dispose() {
    super.dispose();
    WidgetsBinding.instance.removeObserver(this);
  }

  @override
  Widget build(BuildContext context) {
    Consumer _child = Consumer<ThemeController>(builder: (_, provider, __) {
      theme = provider;
      if (widget.automaticSystemBars)
        provider.setSystemBarsColor(
            provider.currentBrightness, widget.statusBarColor);

      return MaterialApp(
        title: this.widget.title,
        home: this.widget.home,
        theme: provider.appTheme,
        darkTheme: provider.systemThemeEnabled
            ? provider.darkTheme
            : provider.appTheme,
      );
    });

    if (widget.controller != null)
      return ChangeNotifierProvider<ThemeController>.value(
          value: widget.controller, child: _child);
    else
      return ChangeNotifierProvider<ThemeController>(
          create: (_) => ThemeController(autoInit: true), child: _child);
  }
}

class DynamicTheming {

  static Future<ThemeController> init() async {
    return await ThemeController(autoInit: false).initProvider();
  }

  static ThemeController of(BuildContext context, { bool updateOnChange = true }) {
    assert(context != null);
    ThemeController result = Provider.of(context, listen: updateOnChange);
    if (result != null) return result;
    throw FlutterError(
      'DynamicTheming.of(context) called with a context that does not implement dynamic theming.'
    );
  }
}


lib/src/theme_provider.dart
import 'package:flutter/material.dart';

// Data Manager
import 'package:dynamic_theming/src/theme_preferences.dart';

// Themes
import 'package:dynamic_theming/src/themes.dart';
import 'package:flutter/services.dart';

class ThemeProvider extends ChangeNotifier {
  ThemePreferences preferences;

  ThemeProvider() {
    initProvider();
  }

  void initProvider() async {
    preferences = await ThemePreferences().init();
    await loadSavedData();
  }

  Color _desaturateColor(Color value) {
    HSLColor color = HSLColor.fromColor(value);
    return HSLColor.fromAHSL(
      (color.alpha),
      (color.hue - 0.449197861),
      (color.saturation - 0.1129032258),
      (color.lightness + 0.1235294118),
    ).toColor();
  }

  void setSystemBarsColor(Brightness platformBrightness, Color statusBarColor) {
    Brightness _statusBrightness = platformBrightness == Brightness.light ?
          Brightness.dark : Brightness.light;
    _statusBarColor = statusBarColor;
    SystemChrome.setSystemUIOverlayStyle(SystemUiOverlayStyle(
      statusBarColor: statusBarColor,
      statusBarBrightness: _statusBrightness,
      statusBarIconBrightness: _statusBrightness,
      systemNavigationBarIconBrightness: platformBrightness,
      systemNavigationBarColor: platformBrightness == Brightness.light ?
          appTheme.scaffoldBackgroundColor :
          darkTheme.scaffoldBackgroundColor,
    ));
  }

  Color _statusBarColor;

  Color _accentColor = Colors.blueAccent;
  Color _darkAccent = Colors.lightBlueAccent;
  bool _systemThemeAvailable = false;
  bool _systemThemeEnabled = false;
  bool _darkThemeEnabled = false;
  bool _blackThemeEnabled = false;

  Color get accentColor => _accentColor;
  Color get darkAccent => _darkAccent;

  bool get systemThemeAvailable => _systemThemeAvailable;
  bool get systemThemeEnabled => _systemThemeEnabled;
  bool get darkThemeEnabled => _darkThemeEnabled;
  bool get blackThemeEnabled => _blackThemeEnabled;

  ThemeData get appTheme {
    if (!darkThemeEnabled || systemThemeEnabled)
      return AppTheme.white(accentColor);
    else
      return darkTheme;
  }

  ThemeData get darkTheme {
    if (blackThemeEnabled)
      return AppTheme.black(darkAccent);
    else
      return AppTheme.dark(darkAccent);
  }

  Brightness get currentBrightness {
    if (systemThemeEnabled)
      return WidgetsBinding.instance.window.platformBrightness;
    else {
      if (darkThemeEnabled)
        return Brightness.dark;
      else
        return Brightness.light;
    }
  }

  set systemThemeAvailable(bool value){
    _systemThemeAvailable = value;
    if (value)
      _systemThemeEnabled = preferences.systemThemeEnabled;
    else 
      _systemThemeEnabled = false;
    notifyListeners();
  }

  set accentColor(Color value) {
    _accentColor = value;
    _darkAccent = _desaturateColor(_accentColor);
    preferences.accentColor = _accentColor;
    notifyListeners();
  }

  set systemThemeEnabled(bool value) {
    _systemThemeEnabled = value;
    preferences.systemThemeEnabled = _systemThemeEnabled;
    setSystemBarsColor(currentBrightness, _statusBarColor);
    notifyListeners();
  }

  set darkThemeEnabled(bool value) {
    _darkThemeEnabled = value;
    preferences.darkThemeEnabled = value;
    setSystemBarsColor(currentBrightness, _statusBarColor);
    notifyListeners();
  }

  set blackThemeEnabled(bool value) {
    _blackThemeEnabled = value;
    preferences.blackThemeEnabled = value;
    setSystemBarsColor(currentBrightness, _statusBarColor);
    notifyListeners();
  }

  Future<void> loadSavedData() async {
    systemThemeAvailable = await preferences.isSystemThemeAvailable();
    accentColor = preferences.accentColor;
    darkThemeEnabled = preferences.darkThemeEnabled;
    blackThemeEnabled = preferences.blackThemeEnabled;
  }
}


lib/src/theme_preferences.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:device_info/device_info.dart';
import 'package:shared_preferences/shared_preferences.dart';

// Key identifiers for variables saved in SharedPreferences.
String _accentKey = "accent_color";
String _systemThemeKey = "use_system_theme";
String _darkThemeKey = "use_dark_theme";
String _blackThemeKey = "use_black_theme";

class ThemePreferences {
  SharedPreferences prefs;

  Future<ThemePreferences> init() async {
    prefs = await SharedPreferences.getInstance();
    return this;
  }

  Future<bool> isSystemThemeAvailable() async {
    String version;
    if(Platform.isAndroid){
      AndroidDeviceInfo deviceInfo = await DeviceInfoPlugin().androidInfo;
      version = deviceInfo.version.release;
    } else if (Platform.isIOS) {
      IosDeviceInfo deviceInfo = await DeviceInfoPlugin().iosInfo;
      version = deviceInfo.systemVersion;
    }
    if (double.parse(version) >= 9 || double.parse(version) >= 13)
      return true;
    else return false;
  }

  Color get accentColor => Color(prefs.getInt(_accentKey) ?? Colors.blueAccent.value);
  set accentColor(Color color) => prefs.setInt(_accentKey, color.value);

  bool get systemThemeEnabled => prefs.getBool(_systemThemeKey) ?? true;
  set systemThemeEnabled(bool value) => prefs.setBool(_systemThemeKey, value);
  
  bool get darkThemeEnabled => prefs.getBool(_darkThemeKey) ?? false;
  set darkThemeEnabled(bool value) => prefs.setBool(_darkThemeKey, value);

  bool get blackThemeEnabled => prefs.getBool(_blackThemeKey) ?? false;
  set blackThemeEnabled(bool value) => prefs.setBool(_blackThemeKey, value);
}


lib/src/theme_controller.dart
import 'package:flutter/material.dart';

// Data Manager
import 'package:dynamic_theming/src/theme_preferences.dart';

// Themes
import 'package:dynamic_theming/src/themes.dart';
import 'package:flutter/services.dart';

class ThemeController extends ChangeNotifier {
  ThemePreferences preferences;

  ThemeController({autoInit = false}) {
    if (autoInit)
      initProvider();
  }

  Future<ThemeController> initProvider() async {
    preferences = await ThemePreferences().init();
    await loadSavedData();
    return this;
  }

  Color _desaturateColor(Color value) {
    HSLColor color = HSLColor.fromColor(value);
    return HSLColor.fromAHSL(
      (color.alpha),
      (color.hue - 0.449197861),
      (color.saturation - 0.1129032258),
      (color.lightness + 0.1235294118),
    ).toColor();
  }

  void setSystemBarsColor(Brightness platformBrightness, Color statusBarColor) {
    Brightness _statusBrightness = platformBrightness == Brightness.light ?
          Brightness.dark : Brightness.light;
    _statusBarColor = statusBarColor;
    SystemChrome.setSystemUIOverlayStyle(SystemUiOverlayStyle(
      statusBarColor: statusBarColor,
      statusBarBrightness: _statusBrightness,
      statusBarIconBrightness: _statusBrightness,
      systemNavigationBarIconBrightness: platformBrightness,
      systemNavigationBarColor: platformBrightness == Brightness.light ?
          appTheme.scaffoldBackgroundColor :
          darkTheme.scaffoldBackgroundColor,
    ));
  }

  Color _statusBarColor;

  Color _accentColor = Colors.blueAccent;
  Color _darkAccent = Colors.lightBlueAccent;
  bool _systemThemeAvailable = false;
  bool _systemThemeEnabled = false;
  bool _darkThemeEnabled = false;
  bool _blackThemeEnabled = false;

  Color get accentColor => _accentColor;
  Color get darkAccent => _darkAccent;

  bool get systemThemeAvailable => _systemThemeAvailable;
  bool get systemThemeEnabled => _systemThemeEnabled;
  bool get darkThemeEnabled => _darkThemeEnabled;
  bool get blackThemeEnabled => _blackThemeEnabled;

  ThemeData get appTheme {
    if (!darkThemeEnabled || systemThemeEnabled)
      return AppTheme.white(accentColor);
    else
      return darkTheme;
  }

  ThemeData get darkTheme {
    if (blackThemeEnabled)
      return AppTheme.black(darkAccent);
    else
      return AppTheme.dark(darkAccent);
  }

  Brightness get currentBrightness {
    if (systemThemeEnabled)
      return WidgetsBinding.instance.window.platformBrightness;
    else {
      if (darkThemeEnabled)
        return Brightness.dark;
      else
        return Brightness.light;
    }
  }

  set systemThemeAvailable(bool value){
    _systemThemeAvailable = value;
    if (value)
      _systemThemeEnabled = preferences.systemThemeEnabled;
    else 
      _systemThemeEnabled = false;
    notifyListeners();
  }

  set accentColor(Color value) {
    _accentColor = value;
    _darkAccent = _desaturateColor(_accentColor);
    preferences.accentColor = _accentColor;
    notifyListeners();
  }

  set systemThemeEnabled(bool value) {
    _systemThemeEnabled = value;
    preferences.systemThemeEnabled = _systemThemeEnabled;
    setSystemBarsColor(currentBrightness, _statusBarColor);
    notifyListeners();
  }

  set darkThemeEnabled(bool value) {
    _darkThemeEnabled = value;
    preferences.darkThemeEnabled = value;
    setSystemBarsColor(currentBrightness, _statusBarColor);
    notifyListeners();
  }

  set blackThemeEnabled(bool value) {
    _blackThemeEnabled = value;
    preferences.blackThemeEnabled = value;
    setSystemBarsColor(currentBrightness, _statusBarColor);
    notifyListeners();
  }

  Future<void> loadSavedData() async {
    systemThemeAvailable = await preferences.isSystemThemeAvailable();
    accentColor = preferences.accentColor;
    darkThemeEnabled = preferences.darkThemeEnabled;
    blackThemeEnabled = preferences.blackThemeEnabled;
  }
}
